monai/apps/pathology/transforms/post/array.py

```diff
@@ -0,0 +1,307 @@
+
+        if image.shape[0] != 1 or image.ndim != 3:
+            raise ValueError(f"Input image should be with size of [1, H, W], but got {image.shape}")
+        if mask.shape[0] != 1 or mask.ndim != 3:
+            raise ValueError(f"Input mask should be with size of [1, H, W], but got {mask.shape}")
+        if markers is not None and (markers.shape[0] != 1 or markers.ndim != 3):
+            raise ValueError(f"Input markers should be with size of [1, H, W], but got {markers.shape}")
+
+        instance_seg = watershed(image, markers=markers, mask=mask, connectivity=self.connectivity)
+
+        return convert_to_dst_type(instance_seg, image, dtype=np.uint8)[0]
+
+
+class GenerateMask(Transform):
+    """
+    generate mask used in `watershed`. Only points at which mask == True will be labeled.
+
+    Args:
+        softmax: if True, apply a softmax function to the prediction.
+        sigmoid: if True, apply a sigmoid function to the prediction.
+        threshold: if not None, threshold the float values to int number 0 or 1 with specified theashold.
+        remove_small_objects: whether need to remove some objects in the marker. Defaults to True.
+        min_size: objects smaller than this size are removed if `remove_small_objects` is True. Defaults to 10.
+
+    Raises:
+        ValueError: when the `prob_map` shape is not [C, H, W].
+
+    """
+
+    backend = [TransformBackends.NUMPY]
+
+    def __init__(
+        self,
+        softmax: bool = True,
+        sigmoid: bool = False,
+        threshold: Optional[float] = None,
+        remove_small_objects: bool = True,
+        min_size: int = 10,
+    ) -> None:
+        if sigmoid and threshold is None:
+            raise ValueError("Threshold is needed when using sigmoid activation.")
+
+        self.activations = Activations(sigmoid=sigmoid, softmax=softmax)
+        self.asdiscrete = AsDiscrete(threshold=threshold, argmax=softmax)
+        if remove_small_objects:
+            self.remove_small_objects = RemoveSmallObjects(min_size=min_size)
+        else:
+            self.remove_small_objects = None  # type: ignore
+
+    def __call__(self, prob_map: NdarrayOrTensor) -> NdarrayOrTensor:
+        """
+        Args:
+            prob_map: probability map of segmentation, shape must be [C, H, W]
+        """
+        if len(prob_map.shape) != 3:
+            raise ValueError(
+                f"Suppose the input probability map should be with shape of [C, H, W], but got {prob_map.shape}"
+            )
+
+        pred = self.activations(prob_map)
+        pred = self.asdiscrete(pred)
+
+        if isinstance(pred, torch.Tensor):
+            pred = pred.detach().cpu()
+
+        pred = label(pred)[0]
+        if self.remove_small_objects:
+            pred = self.remove_small_objects(pred)
+        pred[pred > 0] = 1  # type: ignore
+
+        return convert_to_dst_type(pred, prob_map, dtype=np.uint8)[0]
+
+
+class GenerateProbabilityMap(Transform):
```

Comment:
Sorry, I am a bit confused about this probability map here. Isn't it just a measure that has high values near the borders?
The interior part of a segmentation here has negative value before removing them by ` overall[overall < 0] = 0`, right?

Reply:
Hi @drbeh, you are right, it is a measure used to see which part of the image is not easily distinguished as foreground to give him a greater probability value.

Reply:
Thanks @KumoLiu, right! That's why the name of probability map might be too general and not accurate. They are values normalized between 0 and 1 but they don't sum to 1.

Reply:
Oh, I see. Actually, what I want to say is the map shows the probability of the foreground, maybe we call it `GenerateUncertainArea`? Do you have some suggestions? Thanks!

Reply:
I would suggest to follow this [implementation](https://github.com/Dana-Farber-AIOS/pathml/blob/d5982f4b84708871e635c3db403dec3887021343/pathml/ml/hovernet.py#L719-L739), which is mostly the same the original but with few difference. You can combine it in you code as follow:

1. Add L247: `distance_map = (1.0 - foreground_prob_map) * mask` to this components (L211) and output non-smooth and non-inverted distance map
2. Rename `GenerateProbabilityMap` to `GenerateBackgroundDistanceMap`
3. Remove L247 from `GenerateDistanceMap` and rename it to `GenerateEnergyLandscape` 
4. To generate Markers in L306-308, follow the other implementation and just threshold the background_distance_map.
5. the rest should be the same (except for variable name changing, docstring update, etc.)

Please let me know if it makes sense to you and if you have any question.